### ISSUE TITLE
Recaptcha Field uses inline JavaScript

### ISSUE THREAD

**Issue by BernhardPosselt** [praekelt/django-recaptcha#101]
When using a CSP which disallows inline JavaScript the RecaptchaField breaks. Disallowing inline JavaScript is IMHO the most important thing that you want to disable with CSP. 

Would it be possible to manually add all the required scripts in the \<head> section?

**Comment by BernhardPosselt** [comment]
Same goes for the custom JS config part btw, this should be served under a special URL I think. What's the section supposed to do btw?

**Comment by 9mido** [comment]
I am using django-csp and I followed this https://developers.google.com/recaptcha/docs/faq:

```
I'm using Content-Security-Policy (CSP) on my website. How can I configure it to work with reCAPTCHA?
We recommend using the nonce-based approach documented with CSP3. Make sure to include your nonce in the reCAPTCHA api.js script tag, and we'll handle the rest.

Note: reCAPTCHA also works with 'strict-dynamic' on browsers that support it.

Alternatively, please add the following values to the directives:

script-src https://www.google.com/recaptcha/, https://www.gstatic.com/recaptcha/
frame-src https://www.google.com/recaptcha/

```

The problem though arises when needing to override the templates that this django-recaptcha app uses to be able to edit the templates that this app uses to be able to insert the nonce into the recaptcha app. Relates to issue #211.

**Comment by twschiller** [comment]
If you're using `django-csp` you can at least isolate the CSP changes with their decorator https://django-csp.readthedocs.io/en/latest/decorators.html.

```
from csp.decorators import csp_update
from allauth.account.views import SignupView

@method_decorator(
    csp_update(
        SCRIPT_SRC_ELEM="'self' 'unsafe-inline' https://www.google.com/recaptcha/ https://www.gstatic.com/recaptcha/",
        FRAME_SRC="'self' https://www.google.com/recaptcha/",
    ),
    name="dispatch",
)
class CaptchaSignupView(SignupView):
    pass
```

**Comment by thibaudcolas** [comment]
Yes please! I don’t think anyone has attempted a PR for this, would _love_ to see this happen. I don’t have much CSP expertise myself – from what I can see of the code, a nonce implementation should be reasonably simple. If anyone has thoughts on the best way forward please share here.

**Comment by physicalit** [comment]
Only the html template files need to be adjusted, what if instead of using inline JavaScript just use a remote file?

**Comment by DragosChirila** [comment]
Hi,

IMHO there is a quick way to make `django-recaptcha` CSP compatible/friendly.

Let me share with you the approach we used for our custom widgets.

1. `django-csp` is configured with  `CSP_INCLUDE_NONCE_IN=['script-src']`
2. replace any `<script>` tags with `<script {% if request.csp_nonce %}nonce="{{ request.csp_nonce }}"{% endif %}>`
3. since we are expecting the request object to be available in the widget's template which is not by default, we need to add some extra code:

- each widget must overwrite `get_context`:
```
    def get_context(self, name, value, attrs):
        context = super().get_context(name, value, attrs)

        context.update({
            'request': getattr(self, 'request', None),
        })

        return context
```
- I've used `getattr` instead of just `self.request` to not break existing integrations.
- the form that is using the widget must overwrite `__init__` and set the `request` as a property of the widget:
```
...
captcha = ReCaptchaField(widget=ReCaptchaV2Invisible)
...

    def __init__(self, *args, **kwargs):
        self.request = kwargs.pop('request')
        super().__init__(*args, **kwargs)
        self.fields['captcha'].widget.request = self.request
```
- finally, the view must implement `get_form_kwargs` to pass the `request` object to the `form`:
```
    def get_form_kwargs(self):
        kwargs = super().get_form_kwargs()
        kwargs.update({
            'request': self.request,
        })
        return kwargs
```

I don't think that the inline script can be move to an external JS file since it is dynamic `{{ widget_uuid }}`, from `js_v2_invisible.html`:
```
...
        document.querySelector('.g-recaptcha[data-widget-uuid="{{ widget_uuid }}"]').closest('form').submit();
...
        var element = document.querySelector('.g-recaptcha[data-widget-uuid="{{ widget_uuid }}"]');
        element.closest('form').addEventListener('submit', verifyCaptcha_{{ widget_uuid}});
...
```

Please let me know your thoughts. Is there a better way?

Kind regards,
Dragos

**Comment by Stormheg** [comment]
As already suggested by @physicalit, another alternative could be to ditch the inline scripts entirely and create separate JS static files for each widget. Passing configuration can happen via data attributes or by using the [`json_script`](https://docs.djangoproject.com/en/4.2/ref/templates/builtins/#json-script) tag provided by Django.


Like you mentioned @alanverresen, the hard part about solution 1 is accessing the current `request` from the form field / widget. It feels like a bit much to impose this as a solution/workaround to developers as it requires they write some boilerplate code to pass the request (or the nonce itself) to the form/field/widget. It is not out of the question, just not preferred.

Not entirely sure about solution 2 and 3, but I assume that would also require a lot of changes and/or special template configuration as getting the nonce relies on having access to the request object one way or another.


---

My preference would be to _not_ need the nonce at all and have this package work in a way that is compatible with CSP out of the box.

**Comment by Stormheg** [comment]
@alanverresen I'd welcome a PR that replaces the inline script usage with vanilla JS files 🙂 

Big ask, but if you are up to task... me and the other maintainers sure appreciate a contribution in this direction! Again, no pressure: only if you feel like doing so.

I don't think this has to be a breaking change as long as all option (passed via widget `attrs`) continue working as expected.

**Comment by devwaseem** [comment]
I've quickly made the solution to work by overriding the widget's rendering method and appending the csp_nonce to the script tags.

Here's how I did it:

```
import re
from django_recaptcha.widgets import ReCaptchaV3

class RecaptchaV3WidgetWithCSP(ReCaptchaV3):
    def render(self, name, value, attrs=None, renderer=None):
        rendered_content: str = super().render(name, value, attrs, renderer)
        csp_nonce = getattr(self, "csp_nonce", None)
        if not csp_nonce:
            return rendered_content
        pattern = re.compile(r"(<script\b[^>]*)(>)")
        rendered_content = pattern.sub(f'\\1 nonce="{csp_nonce}"\\2', rendered_content)
        return rendered_content
```

And in the form:

```
class ContactForm(forms.Form):
    def __init__(self, *args: Any, **kwargs: Any):
        request = kwargs.pop("request", None)
        super().__init__(*args, **kwargs)
        if request:
            self.fields["captcha"].widget.csp_nonce = request.csp_nonce
```

and while building the form, just pass the request:

```
form = ContactForm(request.POST, request=request)
```


I hope this helps

**Comment by pierre-dejoue** [comment]
This is quite a serious issue, and I'm glad to see a pull request (#336) is on-going.

To provide some feedback, I've stopped using django-recaptcha completely on my website, due to this issue. I'll happily come back to using this module when it is compatible with a reasonnably safe CSP.
